File: app/src/modules/admin/api/services/role-service.ts
import { http } from '@/core';
import { Role } from '@/modules/admin/api/entities/role';
import { CreateRole } from '@/modules/admin/api/data-transfer-objects/create-role';
import { UpdateRole } from '@/modules/admin/api/data-transfer-objects/update-role';

export class RoleService {
    async getRoles() {
        const res = await http.get<any[]>('/security/role');
        return res.data.map(d => this._map(d));
    }

    async createRole(create: CreateRole) {
        const res = await http.post('/security/role', create);
        return this._map(res.data);
    }

    async updateRole(update: UpdateRole) {
        const res = await http.put(`/security/role/${update.id}`, update);
        return this._map(res.data);
    }

    async deleteRole(role: Role) {
        await http.delete(`/security/role/${role.id}`);
    }

    _map(data: any) {
        const role = new Role(data.name, data.permissionIds);
        role.id = data.id;

        return role;
    }
}

File: api/src/Domain/Security/Services/RoleService.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using DetailingArsenal.Domain.Users;

namespace DetailingArsenal.Domain.Security {
    public interface IRoleService : IService {
        Task<Role> GetById(Guid id);
        Task<List<Role>> GetAll();
        Task<Role> Create(CreateRole create);
        Task Update(Role role, UpdateRole update);
        Task Delete(Role role);

        Task AddRoleToUser(Role role, User user);
        Task RemoveRoleFromUser(Role role, User user);
    }

    public class RoleService : IRoleService {
        IRoleRepo repo;
        RoleNameUniqueSpecification nameUniqueSpec;

        public RoleService(IRoleRepo repo, RoleNameUniqueSpecification nameUniqueSpec) {
            this.repo = repo;
            this.nameUniqueSpec = nameUniqueSpec;
        }

        public async Task<Role> GetById(Guid id) {
            return await repo.FindById(id) ?? throw new EntityNotFoundException();
        }

        public async Task<List<Role>> GetAll() {
            return await repo.FindAll();
        }

        public async Task<Role> Create(CreateRole create) {
            var r = Role.Create(
                create.Name,
                create.PermissionIds
            );

            await nameUniqueSpec.CheckAndThrow(r);
            await repo.Add(r);

            return r;
        }

        public async Task Update(Role role, UpdateRole update) {
            role.Name = update.Name;
            role.PermissionIds = update.PermissionIds;

            await nameUniqueSpec.CheckAndThrow(role);
            await repo.Update(role);
        }

        public async Task Delete(Role role) {
            await repo.Delete(role);
        }

        public async Task AddRoleToUser(Role role, User user) {
            await repo.AddToUser(user, role);
        }

        public async Task RemoveRoleFromUser(Role role, User user) {
            await repo.RemoveFromUser(user, role);
        }
    }
}